mlflow/utils/process.py

```diff
@@ -68,3 +69,29 @@ def _exec_cmd(
     if throw_on_error and prc.returncode != 0:
         raise ShellCommandException.from_completed_process(prc)
     return prc
+
+
+# A global map storing (function, args_tuple) --> (value, pid)
+_per_process_value_cache_map = {}
+
+
+def cache_return_value_per_process(fn):
+    """
+    A decorator which globally cache the return value of the decorated function.
+    But if current process forked out a new child process, in child process,
+    old cache values are invalidated.
+    """
+
+    @functools.wraps(fn)
+    def wrapped_fn(*args):
+        if (fn, args) in _per_process_value_cache_map:
+            prev_value, prev_pid = _per_process_value_cache_map.get((fn, args))
+            if os.getpid() == prev_pid:
+                return prev_value
+
+        new_value = fn(*args)
+        new_pid = os.getpid()
+        _per_process_value_cache_map[(fn, args)] = (new_value, new_pid)
+        return new_value
+
+    return wrapped_fn
```

Comment:
This addressed the issue mentioned in https://github.com/mlflow/mlflow/pull/5561#discussion_r842384536